File: dodam-system-domain/dodam-domain-rds/src/main/java/b1nd/dodam/domain/rds/wakeupsong/service/WakeupSongService.java
package b1nd.dodam.domain.rds.wakeupsong.service;

import b1nd.dodam.core.util.ZonedDateTimeUtil;
import b1nd.dodam.domain.rds.member.entity.Member;
import b1nd.dodam.domain.rds.wakeupsong.entity.WakeupSong;
import b1nd.dodam.domain.rds.wakeupsong.enumeration.WakeupSongStatus;
import b1nd.dodam.domain.rds.wakeupsong.exception.WakeupSongNotFoundException;
import b1nd.dodam.domain.rds.wakeupsong.repository.WakeupSongRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;

@Service
@RequiredArgsConstructor
public class WakeupSongService {

    private final WakeupSongRepository wakeupSongRepository;

    public List<WakeupSong> getByPlayDate(int year, int month, int day) {
        return wakeupSongRepository.findAllByPlayAt(LocalDate.of(year, month, day));
    }

    public List<WakeupSong> getPendingWakeupSong() {
        return wakeupSongRepository.findAllByStatus(WakeupSongStatus.PENDING);
    }

    public List<WakeupSong> getMyWakeupSong(Member member) {
        return wakeupSongRepository.findAllByMember_IdAndStatus(member.getId(), WakeupSongStatus.PENDING);
    }

    public void saveWakeupSong(WakeupSong wakeupSong){
        wakeupSongRepository.save(wakeupSong);
    }

    public Boolean existsByMemberAndCreatedAt(Member member){
        return wakeupSongRepository.existsByMember_IdAndCreatedAtAfter(member.getId(), ZonedDateTimeUtil.nowToLocalDateTime().minusDays(7));
    }

    public WakeupSong getById(int id){
        return wakeupSongRepository.findById(id)
                .orElseThrow(WakeupSongNotFoundException::new);
    }

    public void delete(WakeupSong wakeupSong){
        wakeupSongRepository.delete(wakeupSong);
    }

}

File: services/service-wakeup-song/src/main/kotlin/com/b1nd/dodamdodam/wakeupsong/domain/wakeupsong/service/WakeupSongService.kt
package com.b1nd.dodamdodam.wakeupsong.domain.wakeupsong.service

import com.b1nd.dodamdodam.wakeupsong.domain.wakeupsong.entity.WakeupSongEntity
import com.b1nd.dodamdodam.wakeupsong.domain.wakeupsong.enumeration.WakeupSongStatus
import com.b1nd.dodamdodam.wakeupsong.domain.wakeupsong.exception.WakeupSongNotFoundException
import com.b1nd.dodamdodam.wakeupsong.domain.wakeupsong.repository.WakeupSongRepository
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.util.UUID

@Service
class WakeupSongService(
    private val wakeupSongRepository: WakeupSongRepository
) {

    fun save(wakeupSong: WakeupSongEntity): WakeupSongEntity =
        wakeupSongRepository.save(wakeupSong)

    fun getById(id: Long): WakeupSongEntity =
        wakeupSongRepository.findById(id).orElseThrow { WakeupSongNotFoundException() }

    fun getByUserId(userId: UUID): List<WakeupSongEntity> =
        wakeupSongRepository.findAllByUserId(userId)

    fun getAllowed(year: Int, month: Int, day: Int): List<WakeupSongEntity> {
        val date = LocalDate.of(year, month, day)
        val start = LocalDateTime.of(date, LocalTime.MIN)
        val end = LocalDateTime.of(date, LocalTime.MAX)
        return wakeupSongRepository.findAllByStatusAndCreatedAtBetween(WakeupSongStatus.ALLOWED, start, end)
    }

    fun getPending(): List<WakeupSongEntity> =
        wakeupSongRepository.findAllByStatus(WakeupSongStatus.PENDING)

    fun delete(wakeupSong: WakeupSongEntity) =
        wakeupSongRepository.delete(wakeupSong)
}
